### PR TITLE
test: [RLOS_2023][WIP] add test for regression weight 

### PR DESCRIPTION
**Pull request by michiboo**
Add constant function test for regression

### CODE REVIEW

File: python/tests/test_regression.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from vw_executor.vw import Vw
+from vw_executor.vw_opts import Grid
+from numpy.testing import assert_allclose
+import pandas as pd
+import numpy as np
+import random
+
+with open("regression1.txt", "w") as f:
```
Comment (ataymano) [inline python/tests/test_regression.py:8]: Let's make this call either from test itself or from fixture.

File: python/tests/test_regression.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    for i in range(10000):
```
Comment (ataymano) [inline python/tests/test_regression.py:9]: Generating constant dataset seems generic enough function with data size/constant/... as parameters

File: python/tests/test_regression.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        x = random.uniform(1, 100)
```
Comment (ataymano) [inline python/tests/test_regression.py:10]: it is not great to have true randomness in tests. We should seed it before every call

File: python/tests/test_regression.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    def assert_job(job):
+        tolerance = 1
+        constant = 5
+        for j in job:
+            assert j.status.value == 3
+            data = j.outputs["--readable_model"]
+            with open(data[0], "r") as f:
+                data = f.readlines()
+            data = [i.strip() for i in data]
+            weight_b = float(data[-1].split(":")[1].split(" ")[0])
```
Comment (ataymano) [inline python/tests/test_regression.py:38]: https://github.com/VowpalWabbit/data-science/blob/b9cadfed0622e99050afa5685d77c33aee8703c5/from_mwt_ds/DataScience/vw_executor/artifacts.py#L220

you can have access to weights as job[0].Model9('--readable_model').weights

File: python/tests/test_regression.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            "#base": ["-P 50000 --preserve_performance_counters --save_resume "],
+            "#reg": ["", "--coin", "--ftrl", "--pistol"],
+        }
+    )
+
+    def assert_job(job):
+        tolerance = 1
+        constant = 5
+        for j in job:
+            assert j.status.value == 3
```
Comment (ataymano) [inline python/tests/test_regression.py:33]: let's use enum here:
https://github.com/VowpalWabbit/data-science/blob/b9cadfed0622e99050afa5685d77c33aee8703c5/from_mwt_ds/DataScience/vw_executor/vw.py#L30

File: python/tests/test_regression.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+def test_regression_weight():
+    options = Grid(
+        {
+            "#base": ["-P 50000 --preserve_performance_counters --save_resume "],
+            "#reg": ["", "--coin", "--ftrl", "--pistol"],
+        }
+    )
+
+    def assert_job(job):
```
Comment (ataymano) [inline python/tests/test_regression.py:29]: This code seems quite generic - "compare weights with expected values with given tolerance", but not reusable at the moment.

File: python/tests/test_regression.py (hunk 1 above)
Comment (ataymano) [inline python/tests/test_regression.py:8]: let's also write all data files in separate folder ('data' or something)

File: python/tests/test_regression.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+with open("regression1.txt", "w") as f:
+    for i in range(10000):
+        x = random.uniform(1, 100)
+        y = 5
+        f.write(f"{y} |f x:{x}\n")
+
+
+def core_test(files, grid, outputs, job_assert):
+    vw = Vw(".vw_cache", reset=True)
```
Comment (ataymano) [inline python/tests/test_regression.py:16]: probably better to add handler=None. I think now it is trying to draw progress bars.

File: python/tests/pytest.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+            "#base": ["-P 50000 --preserve_performance_counters --save_resume "],
+            "#reg": ["", "--coin"]
+        }
+    },
+    {
+        "data_func": "constant_function",
+        "data_func_args": [2000,5,1,100],
+        "assert_func": "assert_weight",
+        "assert_func_args": [[5, 0],1],
+        "grid":   {
```
Comment (ataymano) [inline python/tests/pytest.json:27]: learning_rate is specific argument for sgd ("") - others learners have another hyperparmeters (i.e. ftrl has ftrl_alpha and ftrl_beta).

In python there is a way to use '+'/'*' operator over grids in the following way:
```
grid = Grid(
    {
        '#base': '-P 1000 ...'
    }) * (
        Grid({
            "#lr": "",
            "--learning_rate": [0.01, 0.1, 1]
            }) + Grid({
                '#lr': '--ftrl',
                '--ftrl_alpha': [0.01, 0.1, 1]
                }))
```
Would be good to consider being able to express that kind of things in the future.

File: python/tests/test_helper.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import os
+
+# Get the current directory
+current_dir = os.path.dirname(os.path.abspath(__file__))
+
+def json_to_dict_list(file):
+    res = []
+    with open(current_dir + "/" + file, 'r') as file:
+        # Load the JSON data
+        json_data = json.load(file)
```
Comment (ataymano) [inline python/tests/test_helper.py:13]: looks like just returning json_data should work here

File: python/tests/test_regression.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        job_assert(j, *job_assert_args)
+
+@pytest.mark.parametrize('test_dict', json_to_dict_list("pytest.json"), indirect=True)
+def test_all(test_dict):
+    options = Grid(
+      test_dict['grid'],
+    )
+    data = dynamic_function_call("data_generation", test_dict['data_func'], *test_dict["data_func_args"])
+    assert_job = get_function_object("assert_job", test_dict['assert_func'])
+    core_test(data, options, ["--readable_model", "-p"], assert_job, test_dict['assert_func_args'])
```
Comment (ataymano) [inline python/tests/test_regression.py:41]: Can we move output configuration (readable_model/p/...) to json description

File: python/tests/pytest.json (hunk 8 above)
Comment (ataymano) [reply python/tests/pytest.json:27]: this is just to consider for the future

File: python/tests/pytest.json [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+            "#base": ["-P 50000 --preserve_performance_counters --save_resume "],
+            "#reg": ["", "--coin"]
+        }
+    },
+    {
+        "data_func": "constant_function",
+        "data_func_args": [2000,5,1,100],
+        "assert_func": "assert_weight",
```
Comment (ataymano) [inline python/tests/pytest.json:15]: Please consider supporting list of assertions

File: python/tests/test_regression.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
+from vw_executor.vw import Vw
```
Comment (ataymano) [inline python/tests/test_regression.py:1]: this file seems not related to regression but more of a core infrasturcture.
Better to rename it to "core.py/test.py/..."
